### PR TITLE
Add loading of symbols from PDB files (not integrated)

### PR DESCRIPTION
**Pull request by ronaldfw**
We add a PdbFile class that allows loading symbols from a given PDB
file. This is not yet integrated into the symbol loading flow.

For testing, this adds a basic .dll and .pdb compiled from dllmain.cpp.
The latter is also added in this commit together with instructions how
to build the .dll and .pdb.

Tested: Unit tests added.
Bug: http://b/194497774

### CODE REVIEW

File: src/ObjectUtils/PdbFile.cpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+  llvm::pdb::NativeSession* native_session = static_cast<llvm::pdb::NativeSession*>(session_.get());
+  llvm::pdb::PDBFile& pdb_file = native_session->getPDBFile();
+
+  if (!pdb_file.hasPDBDbiStream()) {
+    return ErrorMessage("PDB file does not have a DBI stream.");
+  }
+
+  auto& debug_info_stream = pdb_file.getPDBDbiStream();
+  const auto& modules = debug_info_stream->modules();
```
Comment (dimitry-) [inline src/ObjectUtils/PdbFile.cpp:99]: nit: maybe mention types explicitly in this case - it is not obvious what modules() returns for example.

File: src/ObjectUtils/PdbFile.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+  for (uint32_t index = 0; index < modules.getModuleCount(); ++index) {
+    auto modi = modules.getModuleDescriptor(index);
+    uint16_t modi_stream_index = modi.getModuleStreamIndex();
+
+    if (modi_stream_index == llvm::pdb::kInvalidStreamIndex) {
+      continue;
+    }
+
+    auto mod_stream_data = pdb_file.createIndexedStream(modi_stream_index);
```
Comment (dimitry-) [inline src/ObjectUtils/PdbFile.cpp:109]: nit: same here with auto...

**Comment by ronaldfw** [comment]
Sorry, hold off with reviewing until I've fixed all issues with the checks.

File: src/ObjectUtils/PdbFileTest.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  ASSERT_THAT(symbols_result, HasNoError());
+
+  auto symbols = std::move(symbols_result.value());
+
+  std::vector<SymbolInfo> symbol_infos(symbols.symbol_infos().begin(),
+                                       symbols.symbol_infos().end());
+  EXPECT_EQ(symbol_infos.size(), 4982);
+
+  SymbolInfo symbol = symbol_infos[0];
+  EXPECT_EQ(symbol.name(), "PrintHelloWorldInternal");
```
Comment (dimitry-) [inline src/ObjectUtils/PdbFileTest.cpp:38]: interesting, I would expect this name to be mangled... does it mean pbd info always returns detangled names?

Comment (ronaldfw) [reply src/ObjectUtils/PdbFileTest.cpp:38]: All other names I get area also demangled, not sure if this is a PDB property or if llvm demangles the name when processing the symbol records (couldn't find anything on this in the code, but could have easily missed it).

File: src/ObjectUtils/PdbFile.cpp (hunk 2 above)
Comment (ronaldfw) [reply src/ObjectUtils/PdbFile.cpp:109]: Done.

File: src/ObjectUtils/PdbFile.cpp (hunk 1 above)
Comment (ronaldfw) [reply src/ObjectUtils/PdbFile.cpp:99]: Done.

File: src/ObjectUtils/PdbFile.cpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    std::copy(&(guid.Guid[0]), &(guid.Guid[kGuidSize - 1]), &result[0]);
+    return result;
+  }
+
+  [[nodiscard]] uint32_t GetAge() const override { return session_->getGlobalScope()->getAge(); }
+
+ private:
+  std::filesystem::path file_path_;
+  std::unique_ptr<llvm::pdb::IPDBSession> session_;
+};
```
[the remark this answers is not in the document]
Comment (ronaldfw) [reply src/ObjectUtils/PdbFile.cpp:81]: Done.

File: src/ObjectUtils/PdbFileTest.cpp (hunk 3 above)
Comment (dimitry-) [reply src/ObjectUtils/PdbFileTest.cpp:38]: Add a test for symbol.demangled_name anyways?

Comment (ronaldfw) [reply src/ObjectUtils/PdbFileTest.cpp:38]: Yes, of course, done.

File: src/ObjectUtils/include/ObjectUtils/PdbFile.h [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+class PdbFile {
+ public:
+  PdbFile() = default;
+  virtual ~PdbFile() = default;
+
+  [[nodiscard]] virtual ErrorMessageOr<orbit_grpc_protos::ModuleSymbols> LoadDebugSymbols() = 0;
+  [[nodiscard]] virtual const std::filesystem::path& GetFilePath() const = 0;
+
+  // GUID and Age are used to match PDB files to objects, similar to build id in the ELF case.
+  [[nodiscard]] virtual const std::array<uint8_t, 16> GetGuid() const = 0;
```
Comment (beckerhe) [inline src/ObjectUtils/include/ObjectUtils/PdbFile.h:25]: Nit: `const` is superfluous. (Same as previous comment)

**Review by beckerhe — APPROVED** [review]
Otherwise LGTM

File: src/ObjectUtils/PdbFile.cpp [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+  [[nodiscard]] ErrorMessageOr<orbit_grpc_protos::ModuleSymbols> LoadDebugSymbols() override;
+  [[nodiscard]] const std::filesystem::path& GetFilePath() const override { return file_path_; }
+
+  [[nodiscard]] const std::array<uint8_t, 16> GetGuid() const override {
+    constexpr int kGuidSize = 16;
+    std::array<uint8_t, kGuidSize> result;
+    auto global_scope = session_->getGlobalScope();
+    const llvm::codeview::GUID& guid = global_scope->getGuid();
+    std::copy(&(guid.Guid[0]), &(guid.Guid[kGuidSize - 1]), &result[0]);
```
[the remark this answers is not in the document]
Comment (ronaldfw) [reply src/ObjectUtils/PdbFile.cpp:75]: No, not intended. My bad, thanks for spotting. 

And, yeah, using std::begin and std::end is much better. Thanks for the suggestion.

File: src/ObjectUtils/PdbFile.cpp [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+};
+
+class PdbFileImpl : public PdbFile {
+ public:
+  PdbFileImpl(std::filesystem::path file_path, std::unique_ptr<llvm::pdb::IPDBSession> session);
+
+  [[nodiscard]] ErrorMessageOr<orbit_grpc_protos::ModuleSymbols> LoadDebugSymbols() override;
+  [[nodiscard]] const std::filesystem::path& GetFilePath() const override { return file_path_; }
+
+  [[nodiscard]] const std::array<uint8_t, 16> GetGuid() const override {
```
[the remark this answers is not in the document]
Comment (ronaldfw) [reply src/ObjectUtils/PdbFile.cpp:70]: Thanks, fixed.

File: src/ObjectUtils/include/ObjectUtils/PdbFile.h (hunk 5 above)
Comment (ronaldfw) [reply src/ObjectUtils/include/ObjectUtils/PdbFile.h:25]: Thanks, fixed.

**Comment by ronaldfw** [comment]
@dimitry- Did you wan to take another look? O/w I'll merge the change.

File: src/ObjectUtils/PdbFile.cpp [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    llvm::codeview::SymbolVisitorCallbackPipeline pipeline;
+    llvm::codeview::SymbolDeserializer deserializer(nullptr,
+                                                    llvm::codeview::CodeViewContainer::Pdb);
+    pipeline.addCallbackToPipeline(deserializer);
+    SymbolInfoVisitor symbol_visitor(&module_symbols);
+    pipeline.addCallbackToPipeline(symbol_visitor);
+    llvm::codeview::CVSymbolVisitor visitor(pipeline);
+
+    auto symbol_stream = mod_debug_stream.getSymbolsSubstream();
+    auto& symbol_array = mod_debug_stream.getSymbolArray();
```
Comment (dimitry-) [inline src/ObjectUtils/PdbFile.cpp:136]: nit: explicit type?

**Review by dimitry- — APPROVED** [review]
LGTM

File: src/ObjectUtils/PdbFile.cpp (hunk 8 above)
Comment (ronaldfw) [reply src/ObjectUtils/PdbFile.cpp:136]: Yup, done.

**Comment by ronaldfw** [comment]
Thanks for the reviews, @dimitry- and @beckerhe!

**Pull request merged**